rfc/development-cycle.md

```diff
@@ -0,0 +1,180 @@
+
+
+**Semver check**: Fail (and don't allow merging on this failure) on any PR that doesn't have one, and only one, of the `semver:patch`, `semver:minor`, `semver:major` labels.
+
+**Create backports**: After a PR is merged into the `master` brnach, create backports for that release. It should target all other branches of the same increment according to the "PR Backports" logic above.
+
+
+**Base check**: Verify that the semver label corresponds to the increment of the branch of the base, if it isn't master.
+
+
+**Update/create milestone**: Set the milestone to the next final release of the base branch. Update after every change of base branch and after every commit on the base branch. Create the milestone if it does not exist.
+
+**Do release**:  This should be an action with an API endpoint. It is the most complicated and involved action. It takes place in a number of stages:
+
+1. Input/pre-release check
+2. Bump JS and push release
+3. Create python release and test (optional)
+4. Push python python release
+5. Create new branches and create merge commits (on final release)
+7. Create PRs to other repos for new version (on final release)
+
+It takes three inputs, the branch you are releasing from, whether it's a pre release, and whether to force release without testing. If it is a pre-release, it should be one of `rc`, `beta`, `alpha`.
+
+Before starting it checks to make sure the branch is version branch and that the pre release is valid. You can skip pre release phases, but you cannot go backwards.
+This means if you just did a beta, you cannot do an rc, and if you just did an rc, you cannot do a beta or alpha.
+
+Next, it will checkout the branch in question and create a clean environment. Then it will run the appropriate `jlpm bumpversion` command (these will need to be updated) to create and publish the packages. 
+
+Then it will publish all the JS packages `npm run publish:all`.
+
+Next it will run the end to end tests in a clean environment, if we are not doing a force release, (not the normal unit tests. This let's us publish even if these are failing). If these fail it will push what the JS version changes, but not the python one and exit.
+
+If they succeed, it will upload the wheel to python and push all tags and commits.
+
+Now, if this was a final release, it has some more work to do! First, it will try to merge this release into another branch, as the "Branch Merging" logic dictates above. Then create new branches according to the "Branches" logic above. Push all of that.
+
+Then it will open PRs to the cookiecutter repos, if this was the latest release, to update them.
+
+For any type of release, it will also open a PR against the conda forge repo, to release this version. It will create it on the same branch as the branch we made the release on in JupyterLab.
+
+
+## Open questions
+
+1. How do we deal with changelogs properly? Where do we deploy them from?
```

Comment:
Is there not prior art from matplotlib on this as well?

Reply:
I should ask them... Thinking about how the changelogs works is one of the most confusing parts for me.

I'll spell out how I am thinking about it here, and hopefully I am missing something that will make it simpler.

Let's say I have two branches `1.2.x` and `1.x.0`. I merge a patch into `master`, and this is backported and merged into those two branches.

Now, if I do new patch release next on the `1.2.x`, I assume that changelog entry should show up in that patch release, but *not* in the next minor release.

However, if I do the next minor release first, I should show the changelog entry for the patch PR in that release! And then if I do a patch release later, it should also show it in this as well, I believe?

---

Another changelog question, for each RC or Alpha release, do we make a seperate changelog entry? For the final release, do we show what has just changed since the last pre release, or since the last final release?

---

If we tied changelog entries instead to commit messages or to PR description, instead of to files, we could have more flexibility here in building dynamic changelogs for different scenarios. I could even imagine on the most flexible side, an option where you select two release and it shows you the changelog entries between them.

But this would negatively impact our ability to curate changelogs and merge together multiple PRs into one entry.


@jasongrout you do a lot of changelog work! Maybe you have also have opinions here?

Reply:
We had a good conversation about this with @telamonian  and @vidartf at the weekly meeting:

* When users go to read the docs and click on changelog, they should see all changes, regardless of what branch they were made on.
* doing a "forward port" pr of the added changelog entries on backport releases to the recent releases isn't a crazy way to do this.

I will work on updating this document with these principles in mind. Thank you all for the continued feedback.

Reply:
Thinking about these changelog issues more, I reluctantly am coming to terms with the fact that we might benefit from a richer semantic model of changelog entries. Ignoring the "Which branch are these in??" question at first, I have started thinking about the changelog as a number of entries, each with:

* PR(s): At least one, but maybe multiple PRs which this text summarizes
* Description: Either a single sentence or a richer multiple node (images, paragraphs) description of the feature
* Category (optional): A heading to put this under, like "Developer Changes" or "User Changes" or "Backwards Incompatible extension changes"

I believe this would cover the breadth of our existing changelog, although I would love @jasongrout's input on this since he has been running point on the more in depth ones lately.

Pairing this information with knowledge of what release(s) a certain PR was first present in (maybe multiple b/c of backports), we can create a diff of entries given some number of source releases and some target release. For example, we could ask  "What are all the changes added to 3.0.0 starting from 2.0.0 and 2.2.2?" This would mostly be additions, but there conceptually be some removals, say for entries that were just targeted to some 2.x release that were not included the 3.0.0 release.

Given that, we could create some text description, in Markdown or RST. We could be specific about what the release we are comparing against for each release.

Now for our "Changelog" file in the docs, we would basically run that function a number of times and collect the various responses together.

One open question I have is "What is the base release(s) for some given changelog entry?" For example, for the changelog for 3.0.0, we could say diff from the last prelease of 3.0.0. However, what about for the first pre-release of 3.0.0? From a git perspective, the last release on that branch (the `x.0.0`/`master` branch) would have been `2.0.0`. However, showing all changes since 2.0.0 is not what we are doing currently. What we are doing currently is probably all changes from the last release of 2.x before the first pre-release? 

So some possible version of rules here (these likely need some massaging):

1. *Patch versions* `{x}.{y}.x` branches: For final and pre release patch versions, base the changes off of the previous pre-release/patch. So for say 3.2.1 base it off of the previous release in the 3.2.x branch.
2. *Minor Versions* `{x}.x.0` branches: Base off of last release on this branch. Also, base on last patch release of previous minor release  if that is more recent that latest release on this branch. i.e. a 3.2.0 release should include changes since 3.1.4 not since 3.1.0, if 3.1.4 was released before 3.2.0.
3. * Major versions* `x.0.0` branch: Base off of last release on this branch + last release minor release of previous major release + last patch release of last minor release of previous major release, if any of those are released after last release on this branch.